server/src/test/java/org/elasticsearch/action/admin/cluster/bootstrap/TransportGetDiscoveredNodesActionTests.java

```diff
@@ -289,6 +291,55 @@ public void handleException(TransportException exp) {
 
             assertTrue(countDownLatch.await(10, TimeUnit.SECONDS));
         }
+
+        {
+            final CountDownLatch countDownLatch = new CountDownLatch(1);
+            final GetDiscoveredNodesRequest getDiscoveredNodesRequest = new GetDiscoveredNodesRequest();
+            getDiscoveredNodesRequest.setRequiredNodes(
+                Arrays.asList(localNode.getAddress().toString(), otherNode.getAddress().toString())
+            );
+            getDiscoveredNodesRequest.setTimeout(TimeValue.ZERO);
+            transportService.sendRequest(localNode, GetDiscoveredNodesAction.NAME, getDiscoveredNodesRequest, new ResponseHandler() {
+                @Override
+                public void handleResponse(GetDiscoveredNodesResponse response) {
+                    assertThat(response.getNodes(), containsInAnyOrder(localNode, otherNode));
+                    countDownLatch.countDown();
+                }
+
+                @Override
+                public void handleException(TransportException exp) {
+                    throw new AssertionError("should not be called", exp);
+                }
+            });
+
+            assertTrue(countDownLatch.await(10, TimeUnit.SECONDS));
+        }
+
+        {
+            final CountDownLatch countDownLatch = new CountDownLatch(1);
+            final GetDiscoveredNodesRequest getDiscoveredNodesRequest = new GetDiscoveredNodesRequest();
+            getDiscoveredNodesRequest.setRequiredNodes(
+                Arrays.asList(localNode.getAddress().toString(), localNode.getName())
+            );
+            getDiscoveredNodesRequest.setTimeout(TimeValue.ZERO);
+            transportService.sendRequest(localNode, GetDiscoveredNodesAction.NAME, getDiscoveredNodesRequest, new ResponseHandler() {
+                @Override
+                public void handleResponse(GetDiscoveredNodesResponse response) {
+                    throw new AssertionError("should not be called");
+                }
+
+                @Override
+                public void handleException(TransportException exp) {
+                    Throwable t = exp.getRootCause();
+                    assertThat(t, instanceOf(IllegalArgumentException.class));
+                    assertThat(t.getMessage(), is("Node [" + localNode + "] matched both a name as well as an address entry" +
+                        " in the nodes list specified in setting [cluster.initial_master_nodes]."));
+                    countDownLatch.countDown();
+                }
+            });
+
+            assertTrue(countDownLatch.await(10, TimeUnit.SECONDS));
+        }
```

Comment:
I think we also need (a) a test that matching on node name does work, and (b) a test that matching nonexistent addresses/names does fail.

Reply:
> a test that matching nonexistent addresses/names does fail.

Didn't we decide not to do that in Slack? I remember reading @ywelsch posting something to that effect on Monday?
Either way, in case we want to do this: This means we will actually restrict the allowed master nodes to the given list instead of merely waiting for at least the given nodes as it is implemented now?

Reply:
Ah :) That makes sense => will fix in a sec